### ISSUE TITLE
Warn when sort is called with no arguments on array of non-string types

### ISSUE THREAD

**Issue by fiesh** [eslint/eslint#13098]
**Warn when `Array.prototype.sort()` is called on an array of non-string types without arguments**

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
[5, 0, 10].sort(); // Warn
["5", "0", "10"].sort(); // Don't warn
[5, 0, 10].sort(somePredicate); // Don't warn
```
The reasoning is that this will result in `[0, 10, 5]` since `sort` converts to `string` prior to sorting!  Since three co-workers independently got hit by this, we think this is a mean gotcha of JavaScript.

What you'd want is `[5, 0, 10].sort((lhs, rhs) => lhs - rhs);`

**Why should this rule be included in ESLint (instead of a plugin)?**

Seems like a sensible recommended rule.

**Are you willing to submit a pull request to implement this rule?**

Don't think I'm well-versed enough.

**Comment by kaicataldo** [comment]
While this seems useful, ESLint won't be able to enforce this without a significant number of false negatives, since it doesn't have any type information. As a result, I'm :-1: for inclusion in core.

**Comment by fiesh** [comment]
> While this seems useful, ESLint won't be able to enforce this without a significant number of false negatives, since it doesn't have any type information. As a result, I'm -1 for inclusion in core.

Hmm I see, this is true.  So I suppose the correct plugin for this to be included in would be `typescript-eslint`?

**Comment by anikethsaha** [comment]
I think it should be refactored to a warning that whenever a number is used if it is starting with `0` then it should just warn that it will be converted to decimal. not just for an array . in general it should be. 

eg 
```
< 010
> 8
```

**Comment by fiesh** [comment]
> I think it should be refactored to a warning that whenever a number is used if it is starting with `0` then it should just warn that it will be converted to decimal. not just for an array . in general it should be.
> 
> eg
> 
> ```
> < 010
> > 8
> ```

Not sure I understand this comment.  This doesn't have anything to do with numbers starting with zero, since `[5, 10].sort();` results in `[10, 5]`.

**Comment by anikethsaha** [comment]
In js, I think whenever numbers start with `0` , they get converted to decimal instead of octal.

so, when it's `010`, it converts to `8` so that might show us un-predictable output even though they are correct as per the specs. 

> _Not sure though_

**Comment by fiesh** [comment]
> In js, I think whenever numbers start with `0` , they get converted to decimal instead of octal.
> 
> so, when it's `010`, it converts to `8` so that might show us un-predictable output even though they are correct as per the specs.
> 
> > _Not sure though_

Any number literal starting with `0` is interpreted as octal (as in most languages that borrow from `C`'s syntax I guess.).  So `010 == 8 == 0x08`.  Not sure how this is relevant to this linter rule though.

**Comment by anikethsaha** [comment]
>  should just warn that it will be converted to decimal. not just for an array . in general it should be.

not related to this rule, as in general I mentioned.

**Comment by kaicataldo** [comment]
> Hmm I see, this is true. So I suppose the correct plugin for this to be included in would be typescript-eslint?

@fiesh Yes, I think so!
